crates/stateless/src/witness_db.rs

```diff
@@ -0,0 +1,139 @@
+//! Provides the [`WitnessDatabase`] type, an implementation of [`reth_revm::Database`]
+//! specifically designed for stateless execution environments.
+
+use alloy_primitives::{keccak256, map::B256Map, Address, B256, U256};
+use alloy_rlp::Decodable;
+use alloy_trie::TrieAccount;
+use reth_errors::ProviderError;
+use reth_revm::{bytecode::Bytecode, state::AccountInfo, Database};
+use reth_trie_sparse::SparseStateTrie;
+use std::collections::HashMap;
+use tracing::trace;
+
+/// An EVM database implementation backed by witness data.
+///
+/// This struct implements the [`reth_revm::Database`] trait, allowing the EVM to execute
+/// transactions using:
+///  - Account and storage slot data provided by a [`reth_trie_sparse::SparseStateTrie`].
+///  - Bytecode and ancestor block hashes provided by in-memory maps.
+///
+/// This is designed for stateless execution scenarios where direct access to a full node's
+/// database is not available or desired.
+#[derive(Debug)]
+pub(crate) struct WitnessDatabase<'a> {
+    /// Map of block numbers to block hashes.
+    /// This is used to service the `BLOCKHASH` opcode.
+    // TODO: use Vec instead -- ancestors should be contiguous
+    // TODO: so we can use the current_block_number and an offset to
+    // TODO: get the block number of a particular ancestor
```

Comment:
would getting rid of extra key hashing matter to the prover?

Reply:
I think it needs to be benchmarked first